lib/wibox/widget/template.lua

```diff
@@ -52,6 +61,46 @@ function template:update(args)
     end
 end
 
+function template:set_template(widget_template)
+    local widget = type(widget_template) == "function" and widget_template()
+        or widget_template
+        or wbase.empty_widget()
+
+    self._private.template = widget
+    self._private.widget = wbase.make_widget_from_value(widget)
+
+    -- We need to connect to these signals to actually redraw the template
+    -- widget when its child needs to.
+    local signals = {
+        "widget::redraw_needed",
+        "widget::layout_changed",
+    }
+    for _, signal in pairs(signals) do
+        self._private.widget:connect_signal(signal, function(...)
+            self:emit_signal(signal, ...)
+        end)
+    end
+
+    self:emit_signal("widget::redraw_needed")
+end
+
+function template:get_widget()
+    return self._private.widget
+end
+
+function template:set_update_callback(update_callback)
+    assert(type(update_callback) == "function" or update_callback == nil)
+
+    self._private.update_callback = update_callback
+end
+
+-- @hidden
+function template:set_update_now(update_now)
+    if update_now then
+        self:update()
+    end
+end
```

Comment:
The name doesn't make sense. `set_[something]` implies a setter for a property, but this just triggers a function call.
And the `update_now` argument exists only to call `self:update()` in the constructor, so the entire method could just be replaced with an `if` in the constructor:

```diff
-    ret:set_update_now(args.update_now)
+    if args.update_now then
+        ret:update()
+    end
```

Reply:
This method is required to make the declarative pattern work as expected.

When creating a widget with the declarative pattern, the constructor doesn't receive all the properties as parameters. Instead, the properties are updated after the widget instantiation.

```lua
local w = wibox.widget {
   update_callback = update_fct,
   update_now = true,
   widget = widget.template,
}
```

The  `update_now = true` property is "transformed" to the invocation of the `template:set_update_now` method.

Note that this is why the whole method is documented as a "Hack" and is `@hidden`.

Reply:
Then it would make more sense to just call it `update_now()`, which `drill` [also checks for](https://github.com/awesomeWM/awesome/blob/17eee18d8bd500ac6dc4907701c5ccdb71174c0c/lib/wibox/widget/base.lua#L763-L764).

Reply:
From a clean code PoV, maybe, but at the same time adding too many `if`s to `drill()` might not be super scalable. Another of my PR has some code to make meta-extending the DOM easier using some `_` prefixed extender "I am special, call this and I will do my special stuff" functions. That might be the solution on the longer term.

However for now I think the hack proposed here is good enough. Unexpected, but not in a way which impact end users.

In the even longer term, which is dubious, there is also that issue which calls to replace how the DOM is created and move it to the hierarchy because in practice right now, you can't merge 2 DOMs and that makes `:get_children_by_id()` unreliable when some composed widgets are used... Including this very widget.... This would be an opportunity to revisit some of the bolt-on hacks I made to restore the declarative widget system. Historically, AwesomeWM 3.1-3.4 had it, but 3.5 had a fully rewritten (and much, much better) widget system. It was fully imperative, which annoyed a large number of users who could not port their config. The code was also very heavy for something like `rc.lua`, which is supposed to appear like a config file. Back then I mostly did modules rather than contribute to the core, so I wrote the declarative thing as [the retrograde module](https://github.com/Elv13/retrograde). It eventually got reworked/merged into the core, but was always a bolted on hack rather than something engineered alongside `wibox.hierarchy`. So it would benefit from a larger rework to be more flexible/integrated.

Reply:
> From a clean code PoV, maybe, but at the same time adding too many ifs to drill() might not be super scalable.

I'm not sure to follow what you mean. This `if` branch already exists. 

I have to agree with @sclu1034. A setter is unneeded here. I have somehow missed this feature of `drill`, otherwise I would have used it instead.